Allow installation directory layout to be configured

paulhopkins opened the pull request:
Allow installation directory layout to be configured using either hash_length parameter or spec formatting. Uses one or two configuration options:

```yaml
config:
    install_hash_length: 20
```
or e.g.

```yaml
config:
    install_path_scheme: '${PACKAGE}/${VERSION}/${HASH:20}'
```
Comments welcome

tgamblin commented:
Question: would #2168 work better for you than this option?  That PR allows potentially several custom trees to be created, and the real install layout stays the same, so you could switch layouts if you end up needing to change.  With this PR, you change the actual install layout, so you can get conflicts, and if you want a different layout you'll need to rebuild all the packages.

Do you need to change the install directory layout directly or will links do?

paulhopkins commented:
My primary motivation was to reduce the length of shebangs in user's scripts that access Spack installed packages. Ideally this sort of change should be done at Spack installation only. I don't specifically need the full path editing, I added it for completeness, but it may be useful for people who have to install in already deeply nested locations. (I am also looking at installing on a system where the Spack root would be at least 50 chars.)

> Do you need to change the install directory layout directly or will links do?
Not sure, I will have to look closer at this functionality. As well as exposing the default installations via modules, I am also creating "views" to combine a large set of packages into a single installation directory using hardlinks. The tree command sounds like it is creating a new hierarchy not a combined view so I would need to maintain three ways of accessing the same installations.

alalazo commented:
@paulhopkins What is the behavior when setting `install_hash_length` to a low number causes a directory name clash?

paulhopkins commented:
If I set it hash_length to zero so I get an immediate clash then Spack reports that it finds a spec file with a different hash. At this point you will need to uninstall m4 to continue with a new installation.

```bash
$ spack install m4~sigsegv
==> m4 is already installed in ..../opt/spack/linux-ubuntu16-x86_64/gcc-5.4.0/m4-1.4.17-
$ spack install m4+sigsegv
==> Error: Spec file in ...../opt/spack/linux-ubuntu16-x86_64/gcc-5.4.0/m4-1.4.17-/.spack/spec.yaml does not match hash!
```

citibeth commented:
> My primary motivation was to reduce the length of shebangs in user's scripts that access Spack installed packages.

Have you considered creating a Spack view containing the packages that need to go in the shebangs?  See: http://spack.readthedocs.io/en/latest/workflows.html?highlight=spack%20view

paulhopkins commented:
> Have you considered creating a Spack view containing the packages that need to go in the shebangs?

I already create views of my entire software stack (130 packages) that can alleviate the shebang problem, however, for the time being it makes sense to keep these inside the same directory layout e.g. `linux-ubuntu16-x86_64/gcc-5.4.0/mega-package-abcdef1` that the other packages are installed in.

I would also like to keep the individual packages and modules available for users that want to choose individual packages, or overlay their view with an alternative version of a single package.

At the moment I limit my hash to 20 characters and expect that due to the rapid development of Spack I will  start a new Spack installation long before I get any clashes.

alalazo commented:
> If I set it hash_length to zero so I get an immediate clash then Spack reports that it finds a spec file with a different hash. At this point you will need to uninstall m4 to continue with a new installation

Correct me if I am wrong, but what I think I need is to:

- uninstall m4 *and everything that depends on it* if I need to switch the flavor of `m4`
- *uninstall everything* if for some reason I need both versions of `m4`, because the setting you propose is global and affects packages that are unrelated to `m4`

If I understood this correctly, then I think this can be quite inconvenient.  #2168 on the other hand will just project on a simplified layout and thus you will be sure that there won't be a clash in the real installation tree.

paulhopkins commented:
Granted, but the demo was at the very extreme end of the scale where only
one variant or dependency tree can be installed at once. However, even in
this example multiple versions could be installed, as the version is still
in the path. If you prefer we could set a minimum on this parameter,
perhaps 7 or more. I think that this is a simple configuration parameter
for admins or developers who understand the limitations of their system and
are able to plan ahead. We still keep all the features that Spack has
already developed e.g. smart package layout, environment module support. I
think #2168 is promising, but I don't know if it is in the same state as
views: They can be created and managed, but there is no support for
environment modules, inter-package dependencies can break, and new packages
have to be kept in sync manually.





On 10 January 2017 at 14:41, Massimiliano Culpo <notifications@github.com>
wrote:

> If I set it hash_length to zero so I get an immediate clash then Spack
> reports that it finds a spec file with a different hash. At this point you
> will need to uninstall m4 to continue with a new installation
>
> Correct me if I am wrong, but what I think I need is to:
>
>    - uninstall m4 *and everything that depends on it* if I need to switch
>    the flavor of m4
>    - *uninstall everything* if for some reason I need both versions of m4,
>    because the setting you propose is global and affects packages that are
>    unrelated to m4
>
> If I understood this correctly, then I think this can be quite
> inconvenient. #2168 <https://github.com/LLNL/spack/pull/2168> on the
> other hand will just project on a simplified layout and thus you will be
> sure that there won't be a clash in the real installation tree.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/2789#issuecomment-271592213>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC4vA8w_APbanH23ugHLb_zuPmIszkqlks5rQ5iNgaJpZM4LfO5B>
> .
>



-- 
Paul Hopkins
Computational Infrastructure Scientist
Cardiff University

HopkinsP@cardiff.ac.uk
Office: +44 (0)2920 876964

paulhopkins commented:
@tgamblin and @alalazo mentioned that #2168 may solve my shebang issue as well as providing other features, for example creating multiple trees. I have now properly tested  #2168 and found the following issues:

1. It is highly configurable, but I found the command and the configuration files confusing.
1. There are no commands for displaying the current state of trees.
1. The command can create a new filesystem tree, but does not handle the creation of new environment modules which is my primary method for exposing Spack installations to users.
1. The command must be run every-time a new package is installed so that it is incorporated into the tree. 

I believe that this PR presents much simpler, though much more restrictive, configuration options that retains that core Spack features. It does have drawbacks including a possible increased risk of clashes, however, unless you choose extremely small hash lengths this is still highly unlikely.

scheibelp commented:
1-4 above are definitely issues with #2168 at the moment. However perhaps the most important issue was mentioned earlier:

>> inter-package dependencies can break

To be clear, this was a design choice: when there is a conflict for a prefix #2168 simply picks one of the contenders and slots it in. This is because it's primarily intended to provide a navigation view instead of paths that users should build against (although using them for PATH would be fine). This fits the notion of this PR that hashes (and the properties they provide) are still desired but that shorter ones are expected to be sufficient.

That being said:

>> My primary motivation was to reduce the length of shebangs in user's scripts that access Spack installed packages.

I'm curious why this appeals over an approach like wrapper scripts.

paulhopkins commented:
> I'm curious why shortened prefixes appeals over an approach like wrapper scripts.
When a user first alerted me to this problem I noticed that the ability to change the installation hash length was already in the code (just not exposed as a configuration parameter). I realised that once the change was made, everything else would still continue to work, in particular module generation. I think wrapper scripts or links add another layer of complication, which can lead to other problems.

> hashes (and the properties they provide) are still desired but that shorter ones are expected to be sufficient.
Exactly! Normally, I use 20 characters for the installation hash, but even if I used 9 characters I would need to install a package ~2300 times with the same version but different hashes to get a path collision probability similar to winning the lottery.

mamelara commented:
Hi! We're interested in this branch for our NERSC installs and wondering about the status of this PR.

Also I noticed that there is a bug with ${HASH} spec format and also ${ARCHITECTURE}. A space is added after the architecture spec and also ${HASH} adds an additional forward slash. So we end up getting directory trees looking like /platform-os-target //hash.

paulhopkins commented:
@mamelera There is are issues with this PR that @tgamblin said would need to be fixed before this can be accepted. From the [telecon minutes](https://github.com/LLNL/spack/wiki/Telcon%3A-2017-02-23):

- [ ] We'll need to take this into account when we integrate binary packaging
not only re-RPATH things but also adjust the install location to the new directory layout
- [ ] install layout should probably be a function of the particular package tree - spack currently only supports one install root, but we'd like to support multiple (one spack relying on another)
- [ ] description of the install layout should be stored with the particular root, not in config.yaml, spack will need to look at description of the install tree to figure out what to do
- [ ] Spack will need to bail out gracefully when there is a hash conflict due to a short install location.

Also I have found that 
- [ ] PR #2941 changed the formatting of the `$HASH` and `$ARCHITECTURE` which does break installation paths.
- [ ] If the directory layout depth is changed from the default then the `spack reindex` command will fail to find these directories.

I will try and fix these issues and re-submit

mamelara commented:
@paulhopkins thank you for the update. Please let me know if you need any help or testing. We are currently relying on this PR for our installs. Luckily we haven't moved into production yet but we're hoping to start relatively soon.

scheibelp commented:
Hi Paul,

@tgamblin has given me the authority for the ultimate OK on this so I will be going over this today and 4/3 (and I expect to be finished by the end of 4/3)

I'd say that in short things look generally good to me and I don't think all the constraints established at the telecon need to be satisfied in this PR; I have the following comments so far:

> We'll need to take this into account when we integrate binary packaging
not only re-RPATH things but also adjust the install location to the new directory layout

> install layout should probably be a function of the particular package tree - spack currently only supports one install root, but we'd like to support multiple (one spack relying on another)

> description of the install layout should be stored with the particular root, not in config.yaml, spack will need to look at description of the install tree to figure out what to do

While someone may have to modify your code to support these concepts, IMO your PR doesn't make it much more difficult to do so when it is decided to proceed with this. Therefore I don't think you need to make any special accommodations. From looking at the binary packages PR (https://github.com/LLNL/spack/pull/445) it appears that only minor modifications to ```relocate_package``` would be needed. Heads up to @hegner in case you disagree (apologies as it appears your PR has been around for a while).

> Spack will need to bail out gracefully when there is a hash conflict due to a short install location.

Spack should already do this (part of ```directory_layout.check_installed```). Perhaps it would be worthwhile to add a test to ensure this occurs (e.g. if you set hash length to 0) but I don't consider that necessary.

> PR #2941 changed the formatting of the $HASH and $ARCHITECTURE which does break installation paths.

IMO Spec.format should just print the hash without a preceding "/" when the user specifies ```{HASH}```. This would be consistent with the normal distinction between the format shorthand sequences like ```@``` vs. ```version```. @becker33 is there a reason that both ```{HASH}``` and its shorthand prefix a "/"?

> If the directory layout depth is changed from the default then the spack reindex command will fail to find these directories.

I'm not sure what you mean when you say "directory layout depth". I could see how ```spack reindex``` would "fail" if anything about the path scheme changes, although IMO that would be fine since the path scheme is the definition of where things are installed. In other words I'm fine with the notion that changing the path scheme invalidates all current installs.

scheibelp commented on lib/spack/spack/schema/config.py:
```diff
@@ -41,6 +41,8 @@
             'additionalProperties': False,
             'properties': {
                 'install_tree': {'type': 'string'},
+                'install_hash_length': {'type': 'integer', 'minimum': 0},
```
Can you set this to be "1" instead of "0"? I think creating a distinction between 0 and >0 is worthwhile. Not using a hash at all increases the collision likelihood to the point where IMO spack would not be usable. Certainly a hash length of 1 would also be pretty bad but forcing the use of __some__ hash forces the user to consider what the length should be.

paulhopkins replied:
Fixed

scheibelp commented on lib/spack/spack/directory_layout.py:
```diff
@@ -150,24 +151,31 @@ def remove_install_directory(self, spec):
-        self.hash_len       = kwargs.get('hash_len', None)
+        self.hash_len       = kwargs.get('hash_len')
+        self.path_scheme    = kwargs.get('path_scheme') or (
+            "${ARCHITECTURE}/${COMPILERNAME}-${COMPILERVER}/${PACKAGE}-${VERSION}-${HASH}")  # NOQA: ignore=E501
```
It was recently discovered that ```NOQA: ignore``` ends up ignoring all flake errors. It seems the correct way to ignore the specific error is ```# NOQA: E501```

paulhopkins replied:
Fixed

scheibelp commented on lib/spack/docs/config_yaml.rst:
```diff
@@ -42,6 +42,39 @@ or with braces to distinguish the variable from surrounding characters:
+comprehend and if very long can also create problems for scripts with
+hardcoded shebangs. There are two parameters to help with that. The
+``install_hash_length`` parameter can set the length of the hash in
+the installation path from 0 to 32. The default uses the full 32 characters.
```
If you agree with my comment about using a minimimum > 1 in the schema for config.yaml, could you update this doc?

paulhopkins replied:
Fixed

paulhopkins commented:
Hi @scheibelp. Thanks for looking at this; 

I have fixed the minor errors you pointed out, including making the minimum hash length 1.

> Spack will need to bail out gracefully when there is a hash conflict due to a short install location.
> Spack should already do this (part of directory_layout.check_installed). Perhaps it would be worthwhile to add a test to ensure this occurs (e.g. if you set hash length to 0) but I don't consider that necessary.

If there is a hash collision Spack already prints the following:
`
==> Error: Spec file in /home/paul/Projects/LIGO/spack/opt/spack/linux-ubuntu16-x86_64/gcc-6.2.0/python-2.7.13-t/.spack/spec.yaml does not match hash!
`

> PR #2941 changed the formatting of the $HASH and $ARCHITECTURE which does break installation paths.
> IMO Spec.format should just print the hash without a preceding "/" when the user specifies {HASH}. This would be consistent with the normal distinction between the format shorthand sequences like @ vs. version. @becker33 is there a reason that both {HASH} and its shorthand prefix a "/"?

I have modified the formatting to not include the `"/"` at the beginning of `${HASH}` and also removed the space at the end of `${ARCHITECTURE}`

> I'm not sure what you mean when you say "directory layout depth".
`spack find` uses a wildcard search to find the `spec.yaml` files, not recursive search. Therefore, I have modified the code so that it uses the appropriate number of wildcards for your custom search path.

> In other words I'm fine with the notion that changing the path scheme invalidates all current installs.
@tgamblin suggested that the path parameters be installed in the installation directory /  database so that they cannot be accidentally changed. Do you want me to commit something before this is merged?

scheibelp commented:
>>In other words I'm fine with the notion that changing the path scheme invalidates all current installs.

>@tgamblin suggested that the path parameters be installed in the installation directory / database so that they cannot be accidentally changed. Do you want me to commit something before this is merged?

If it were stored in the spack install directory (```spack/opt/``` I presume) that would make it less accessible but vulnerable in the same manner.

Storing it in the database file may help but IMO it's not required: install_tree already doesn't have that sort of protection. I think it would be useful add a comment in the documentation to warn the user about the consequences of changing install_tree, install_path_scheme, or install_hash_length.

So in short: not necessary

It looks like the NOQA change didn't make it through and there is a flake issue, but otherwise LGTM.

paulhopkins commented:
Thanks Peter, the requested fixes have been made

scheibelp commented:
Awesome - thanks for your work on this!
